File: mobile/shared/src/commonTest/kotlin/com/storytail/adventures/domain/messages/InboxTest.kt
package com.storytail.adventures.domain.messages

import com.storytail.adventures.api.InboxConversationView
import kotlinx.datetime.LocalDate
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The twin of `web/lib/messages/inbox.test.ts`.
 *
 * `today` is passed explicitly in every case. A relative-time formatter tested against the real
 * clock passes at 3pm and fails at midnight, which is the worst kind of flake — it lands in CI
 * on somebody else's branch.
 *
 * `formatInboxTime`'s clock-time branch is NOT asserted against an exact string here, and that
 * is deliberate rather than lazy: unlike the web twin, which is handed
 * `platform_user.time_zone`, the native side reads the DEVICE zone, so "1:30a" depends on where
 * the machine running the test is. What is asserted is the branch taken — a time for today, a
 * date for anything older — which is the logic this file owns.
 */
class InboxTest {

    private val today = LocalDate(2026, 3, 15)

    private fun conversation(
        id: String = "0195a2c0-1a00-7000-8000-00000000abcd",
        subject: String? = null,
        tripId: String? = null,
        tripTitle: String? = null,
        lastMessageAt: String = "2026-03-15T12:00:00.000Z",
        preview: String? = "Locked. Final balance authorization is in your dashboard.",
        unread: Int = 0,
    ) = InboxConversationView(
        id = id,
        subject = subject,
        tripId = tripId,
        tripTitle = tripTitle,
        lastMessageAt = lastMessageAt,
        lastMessagePreview = preview,
        unreadCount = unread,
    )

    @Test
    fun `names a thread by trip, then subject, then the advisor`() {
        val rows = inboxRows(
            listOf(
                conversation(id = "a", tripTitle = "Sandals · Aug", subject = "Stale old subject"),
                conversation(id = "b", subject = "New family cruise idea"),
                conversation(id = "c"),
            ),
            today,
        )

        // The trip can be renamed after the thread was created, so the header follows the trip.
        assertEquals("Sandals · Aug", rows[0].title)
        assertEquals("New family cruise idea", rows[1].title)
        // The general thread is named for the person, not described as "General".
        assertEquals(MessagesMessages.GENERAL_THREAD_TITLE, rows[2].title)
    }

    @Test
    fun `shows a date rather than a clock time for an older message`() {
        val rows = inboxRows(
            listOf(conversation(lastMessageAt = "2026-03-01T12:00:00.000Z")),
            today,
        )
        assertEquals("Mar 1", rows[0].time)
    }

    @Test
    fun `shows a clock time for today, which is never a date`() {
        val rows = inboxRows(listOf(conversation(lastMessageAt = "2026-03-15T12:00:00.000Z")), today)
        val time = rows[0].time
        // Mid-day UTC is the same calendar day in every zone from UTC-11 to UTC+11, so this is
        // "today" wherever the test runs — and today renders as a clock time, which carries a
        // meridiem suffix and no month name.
        assertTrue(time.endsWith("a") || time.endsWith("p"), "expected a clock time, got '$time'")
        assertTrue(time.contains(":"), "expected a clock time, got '$time'")
    }

    @Test
    fun `renders a row whose preview column is null`() {
        // `last_message_preview` is nullable, and `trip-message` is not atomic — a conversation
        // exists for a moment before its first message lands. A thread that vanishes because
        // one column is null is worse than a thread with a quiet line under it.
        val rows = inboxRows(listOf(conversation(preview = null)), today)
        assertEquals("", rows[0].preview)
        assertEquals(1, rows.size)
    }

    @Test
    fun `returns an empty string rather than throwing on an unusable timestamp`() {
        assertEquals("", formatInboxTime("not a date", today))
    }

    @Test
    fun `carries the unread count through untouched`() {
        assertEquals(3, inboxRows(listOf(conversation(unread = 3)), today)[0].unreadCount)
    }

    @Test
    fun `search matches the title and the preview, ignoring case`() {
        val rows = inboxRows(
            listOf(
                conversation(id = "a", tripTitle = "Sandals · Aug", preview = "Bungalow held"),
                conversation(id = "b", subject = "Aruba honeymoon", preview = "Mid-October?"),
            ),
            today,
        )

        assertEquals(listOf("Sandals · Aug"), filterInboxRows(rows, "SANDALS").map { it.title })
        assertEquals(listOf("Sandals · Aug"), filterInboxRows(rows, "bungalow").map { it.title })
        assertEquals(listOf("Aruba honeymoon"), filterInboxRows(rows, "aruba").map { it.title })
    }

    @Test
    fun `search returns everything for a blank query and nothing for a miss`() {
        val rows = inboxRows(listOf(conversation(id = "a"), conversation(id = "b")), today)
        assertEquals(2, filterInboxRows(rows, "").size)
        assertEquals(2, filterInboxRows(rows, "   ").size)
        // The screen renders SEARCH_EMPTY off this. Returning the unfiltered list on a miss
        // would read as "search is broken" rather than "no results".
        assertEquals(0, filterInboxRows(rows, "zzz").size)
    }
}

File: web/lib/messages/inbox.test.ts
import { describe, expect, it } from "vitest";

import { filterRows, formatInboxTime, inboxRows, type InboxRow } from "./inbox";
import type { InboxConversation } from "./queries";

/**
 * `now` is fixed in every test. A relative-time formatter tested against the real clock
 * passes at 3pm and fails at midnight, which is the worst kind of flake: it lands in CI on
 * somebody else's branch.
 */
const NOW = new Date("2026-03-15T02:00:00.000Z");

function conversation(overrides: Partial<InboxConversation> = {}): InboxConversation {
  return {
    id: "0195a2c0-1a00-7000-8000-00000000abcd",
    subject: null,
    tripId: null,
    tripTitle: null,
    lastMessageAt: "2026-03-15T01:30:00.000Z",
    lastMessagePreview: "Locked. Final balance authorization is in your dashboard.",
    unreadCount: 0,
    ...overrides,
  };
}

describe("formatInboxTime", () => {
  it("shows a clock time for today and a date for older days", () => {
    expect(formatInboxTime("2026-03-15T01:30:00.000Z", "UTC", NOW)).toBe("1:30a");
    expect(formatInboxTime("2026-03-14T23:00:00.000Z", "UTC", NOW)).toBe("Yesterday");
    expect(formatInboxTime("2026-03-01T12:00:00.000Z", "UTC", NOW)).toBe("Mar 1");
  });

  it("decides 'today' in the traveler's zone, not the server's", () => {
    // 23:00Z on Mar 14 is 7:00p on Mar 14 in New York, where `now` is also still Mar 14 —
    // so it is TODAY there and YESTERDAY in UTC. This is the whole reason the zone is an
    // argument: rendered on a UTC host, the ambient-zone version tells a traveler in New
    // York that the message they are reading arrived yesterday.
    const iso = "2026-03-14T23:00:00.000Z";
    expect(formatInboxTime(iso, "America/New_York", NOW)).toBe("7:00p");
    expect(formatInboxTime(iso, "UTC", NOW)).toBe("Yesterday");
  });

  it("returns an empty string rather than throwing on an unusable timestamp", () => {
    expect(formatInboxTime("not a date", "UTC", NOW)).toBe("");
  });

  it("falls back to UTC on a garbage zone instead of blanking the row", () => {
    // `platform_user.time_zone` is free text that onboarding wrote. One bad row should cost
    // a wrong hour, not a thrown render that takes the inbox down with it.
    expect(formatInboxTime("2026-03-15T01:30:00.000Z", "Mars/Olympus_Mons", NOW)).toBe("1:30a");
  });
});

describe("inboxRows", () => {
  it("prefers the live trip title, then the subject, then the advisor's name", () => {
    const rows = inboxRows(
      [
        conversation({ id: "a", tripTitle: "Sandals · Aug", subject: "Stale old subject" }),
        conversation({ id: "b", subject: "New family cruise idea" }),
        conversation({ id: "c" }),
      ],
      "UTC",
      NOW,
    );

    // The trip can be renamed after the thread was created, so the header follows the trip.
    expect(rows[0].title).toBe("Sandals · Aug");
    expect(rows[1].title).toBe("New family cruise idea");
    // The general thread is named for the person, not described as "General".
    expect(rows[2].title).toBe("Gyasi Story");
  });

  it("renders a row whose preview column is null", () => {
    // `last_message_preview` is nullable and a conversation exists for a moment before its
    // first message lands. A thread that vanishes because one column is null is worse than
    // a thread with a quiet line under it.
    const [row] = inboxRows([conversation({ lastMessagePreview: null })], "UTC", NOW);
    expect(row.preview).toBe("");
    expect(row.href).toBe("/messages/0195a2c0-1a00-7000-8000-00000000abcd");
  });

  it("carries the unread count through untouched", () => {
    const [row] = inboxRows([conversation({ unreadCount: 3 })], "UTC", NOW);
    expect(row.unreadCount).toBe(3);
  });
});

describe("filterRows", () => {
  const rows: InboxRow[] = inboxRows(
    [
      conversation({ id: "a", tripTitle: "Sandals · Aug", lastMessagePreview: "Bungalow held" }),
      conversation({ id: "b", subject: "Aruba honeymoon", lastMessagePreview: "Mid-October?" }),
    ],
    "UTC",
    NOW,
  );

  it("returns everything for an empty or whitespace query", () => {
    expect(filterRows(rows, "")).toHaveLength(2);
    expect(filterRows(rows, "   ")).toHaveLength(2);
  });

  it("matches the title and the preview, ignoring case", () => {
    expect(filterRows(rows, "SANDALS").map((r) => r.title)).toEqual(["Sandals · Aug"]);
    expect(filterRows(rows, "bungalow").map((r) => r.title)).toEqual(["Sandals · Aug"]);
    expect(filterRows(rows, "aruba").map((r) => r.title)).toEqual(["Aruba honeymoon"]);
  });

  it("matches nothing rather than everything when nothing matches", () => {
    // The screen renders `searchEmpty` off this. Returning the unfiltered list on a miss
    // would read as "search is broken" rather than "no results".
    expect(filterRows(rows, "zzz")).toEqual([]);
  });
});
